File: backend/orbit8d/engine/hrtf.py
"""HRTF 网格：从 SOFA 加载 → 顺时针方位索引 → 重采样 → 归一；双线性插值；前端用的二进制格式。

约定见 docs/SPEC.md §4.1 与 §5.6。data[el, az, ear, tap]，ear 0 = 左耳。
"""

import json
import struct
from dataclasses import dataclass
from pathlib import Path

import h5py
import numpy as np
from scipy.signal import resample_poly

MAGIC = b"O8DH"
FORMAT_VERSION = 1
TAPS = 128
AZ_TOLERANCE_DEG = 1e-3


@dataclass(frozen=True)
class HrtfGrid:
    sample_rate: int
    az_step_deg: float
    el_nodes: np.ndarray  # (n_el,) 升序，单位度
    data: np.ndarray  # (n_el, n_az, 2, taps) float32

    @property
    def n_el(self) -> int:
        return self.data.shape[0]

    @property
    def n_az(self) -> int:
        return self.data.shape[1]

    @property
    def taps(self) -> int:
        return self.data.shape[3]


def _left_ear_first(receivers: np.ndarray, ir: np.ndarray) -> np.ndarray:
    """SOFA 坐标 +y 指向左侧；保证 ear 0 是左耳。"""
    y = receivers.reshape(receivers.shape[0], -1)[:, 1]
    return ir if y[0] > y[1] else ir[:, ::-1, :]


def from_sofa(path: Path, sample_rate: int) -> HrtfGrid:
    with h5py.File(path, "r") as f:
        ir = np.array(f["Data.IR"], dtype=np.float64)  # (M, 2, N)
        fs = int(round(float(np.array(f["Data.SamplingRate"]).ravel()[0])))
        pos = np.array(f["SourcePosition"], dtype=np.float64)
        receivers = np.array(f["ReceiverPosition"], dtype=np.float64)
    ir = _left_ear_first(receivers, ir)

    el_nodes = np.unique(np.round(pos[:, 1], 6))
    az_ccw = np.mod(np.round(pos[:, 0], 6), 360.0)
    az_cw = np.mod(360.0 - az_ccw, 360.0)
    az_values = np.unique(np.round(az_cw, 6))
    steps = np.diff(az_values)
    if not len(steps) or np.ptp(steps) > AZ_TOLERANCE_DEG or len(az_values) * len(el_nodes) != len(pos):
        raise ValueError(f"{path.name} 不是规整的方位×仰角网格")
    step = float(steps[0])

    el_idx = np.searchsorted(el_nodes, np.round(pos[:, 1], 6))
    az_idx = np.round(az_cw / step).astype(int) % len(az_values)
    grid = np.zeros((len(el_nodes), len(az_values), 2, ir.shape[-1]))
    grid[el_idx, az_idx] = ir

    g = np.gcd(sample_rate, fs)
    grid = resample_poly(grid, sample_rate // g, fs // g, axis=-1)
    grid = (
        grid[..., :TAPS]
        if grid.shape[-1] >= TAPS
        else np.pad(grid, [(0, 0)] * 3 + [(0, TAPS - grid.shape[-1])])
    )

    row = int(np.argmin(np.abs(el_nodes)))
    grid /= np.sqrt((grid[row] ** 2).sum(-1).mean())  # 水平一圈平均每耳能量 = 1
    return HrtfGrid(
        sample_rate=sample_rate,
        az_step_deg=step,
        el_nodes=el_nodes.astype(np.float64),
        data=grid.astype(np.float32),
    )


def interp_weights(grid: HrtfGrid, az: np.ndarray, el: np.ndarray):
    """返回 (i0, i1, j0, j1, t_el, t_az)，用于在 (el, az) 网格上双线性插值。"""
    az = np.mod(np.asarray(az, dtype=np.float64), 360.0)
    el = np.asarray(el, dtype=np.float64)
    nodes = grid.el_nodes
    i0 = np.clip(np.searchsorted(nodes, el, side="right") - 1, 0, grid.n_el - 2)
    i1 = i0 + 1
    t_el = np.clip((el - nodes[i0]) / (nodes[i1] - nodes[i0]), 0.0, 1.0)
    u = az / grid.az_step_deg
    j0f = np.floor(u)
    t_az = u - j0f
    j0 = j0f.astype(int) % grid.n_az
    j1 = (j0 + 1) % grid.n_az
    return i0, i1, j0, j1, t_el, t_az


def blend(table: np.ndarray, weights) -> np.ndarray:
    """table 的前两维是 (el, az)；按插值权重混合，保留其余维度。"""
    i0, i1, j0, j1, t_el, t_az = weights
    shape = (-1,) + (1,) * (table.ndim - 2)
    te, ta = t_el.reshape(shape), t_az.reshape(shape)
    return (
        (1 - te) * (1 - ta) * table[i0, j0]
        + (1 - te) * ta * table[i0, j1]
        + te * (1 - ta) * table[i1, j0]
        + te * ta * table[i1, j1]
    )


def interpolate(grid: HrtfGrid, az, el) -> np.ndarray:
    """(n, 2, taps)：每个方向的左右耳 HRIR。"""
    return blend(grid.data, interp_weights(grid, az, el)).astype(np.float32)


def to_bytes(grid: HrtfGrid) -> bytes:
    header = json.dumps(
        {
            "version": FORMAT_VERSION,
            "sample_rate": grid.sample_rate,
            "taps": grid.taps,
            "az_step_deg": grid.az_step_deg,
            "n_az": grid.n_az,
            "el_nodes": grid.el_nodes.tolist(),
            "layout": "el,az,ear,tap",
        }
    ).encode()
    pad = (-len(header)) % 4  # 数据段 4 字节对齐
    header += b" " * pad
    return MAGIC + struct.pack("<I", len(header)) + header + grid.data.astype("<f4").tobytes()


def from_bytes(blob: bytes) -> HrtfGrid:
    if blob[:4] != MAGIC:
        raise ValueError("不是 Orbit 8D 的 HRTF 数据")
    (hlen,) = struct.unpack("<I", blob[4:8])
    meta = json.loads(blob[8 : 8 + hlen])
    if meta.get("version") != FORMAT_VERSION:
        raise ValueError(f"不支持的 HRTF 数据版本: {meta.get('version')}")
    nodes = np.array(meta["el_nodes"], dtype=np.float64)
    data = np.frombuffer(blob[8 + hlen :], dtype="<f4").reshape(len(nodes), meta["n_az"], 2, meta["taps"])
    return HrtfGrid(
        sample_rate=meta["sample_rate"],
        az_step_deg=meta["az_step_deg"],
        el_nodes=nodes,
        data=data.astype(np.float32),
    )


def load_grid(sofa_path: Path, cache_dir: Path, sample_rate: int) -> HrtfGrid:
    """优先读缓存的二进制表；没有就从 SOFA 生成并缓存（原子写）。"""
    cache = cache_dir / f"hrtf_{sample_rate}.bin"
    if cache.exists():
        return from_bytes(cache.read_bytes())
    grid = from_sofa(sofa_path, sample_rate)
    tmp = cache.with_suffix(".tmp")
    tmp.write_bytes(to_bytes(grid))
    tmp.replace(cache)
    return grid

File: web/src/audio/hrtf.ts
// HRTF 表：解析后端 /api/assets/hrtf.bin（格式见 docs/SPEC.md §5.6），双线性插值与 Python 版一致。

export interface HrtfTable {
  sampleRate: number;
  taps: number;
  azStepDeg: number;
  nAz: number;
  elNodes: Float64Array;
  data: Float32Array; // 布局 [el][az][ear][tap]
}

const MAGIC = "O8DH";
const FORMAT_VERSION = 1;
const HEADER_OFFSET = 8;

interface HrtfHeader {
  version: number;
  sample_rate: number;
  taps: number;
  az_step_deg: number;
  n_az: number;
  el_nodes: number[];
  layout: string;
}

export function parseHrtf(buffer: ArrayBuffer): HrtfTable {
  const bytes = new Uint8Array(buffer);
  if (String.fromCharCode(...bytes.subarray(0, 4)) !== MAGIC) throw new Error("不是 Orbit 8D 的 HRTF 数据");
  const headerLen = new DataView(buffer).getUint32(4, true);
  const header = JSON.parse(new TextDecoder().decode(bytes.subarray(HEADER_OFFSET, HEADER_OFFSET + headerLen))) as HrtfHeader;
  if (header.version !== FORMAT_VERSION || header.layout !== "el,az,ear,tap") {
    throw new Error(`不支持的 HRTF 数据版本: ${header.version}`);
  }
  const count = header.el_nodes.length * header.n_az * 2 * header.taps;
  const data = new Float32Array(buffer, HEADER_OFFSET + headerLen, count);
  return {
    sampleRate: header.sample_rate,
    taps: header.taps,
    azStepDeg: header.az_step_deg,
    nAz: header.n_az,
    elNodes: Float64Array.from(header.el_nodes),
    data,
  };
}

/** 在 (el, az) 网格上双线性插值，写入 outL / outR（长度 = taps）。 */
export function interpolateHrir(t: HrtfTable, azDeg: number, elDeg: number, outL: Float64Array, outR: Float64Array): void {
  const nodes = t.elNodes;
  const nEl = nodes.length;
  let i0 = 0;
  while (i0 < nEl - 2 && nodes[i0 + 1] <= elDeg) i0++;
  const i1 = i0 + 1;
  const tEl = Math.min(1, Math.max(0, (elDeg - nodes[i0]) / (nodes[i1] - nodes[i0])));
  const az = ((azDeg % 360) + 360) % 360;
  const u = az / t.azStepDeg;
  const j0f = Math.floor(u);
  const tAz = u - j0f;
  const j0 = ((j0f % t.nAz) + t.nAz) % t.nAz;
  const j1 = (j0 + 1) % t.nAz;
  const w00 = (1 - tEl) * (1 - tAz);
  const w01 = (1 - tEl) * tAz;
  const w10 = tEl * (1 - tAz);
  const w11 = tEl * tAz;
  const taps = t.taps;
  const stride = 2 * taps;
  const b00 = (i0 * t.nAz + j0) * stride;
  const b01 = (i0 * t.nAz + j1) * stride;
  const b10 = (i1 * t.nAz + j0) * stride;
  const b11 = (i1 * t.nAz + j1) * stride;
  const d = t.data;
  for (let k = 0; k < taps; k++) {
    outL[k] = w00 * d[b00 + k] + w01 * d[b01 + k] + w10 * d[b10 + k] + w11 * d[b11 + k];
    const r = taps + k;
    outR[k] = w00 * d[b00 + r] + w01 * d[b01 + r] + w10 * d[b10 + r] + w11 * d[b11 + r];
  }
}
